File: examples/tracking/multi-camera-people-detection-and-tracking-optiview/cpp/sample_utils.cpp
#include "examples/tracking/multi-camera-people-detection-and-tracking-optiview/cpp/sample_utils_api.cpp"

#include "support/object_detection/obj_detection_utils.h"

#include "pipeline/TensorCore.h"

#include <stdexcept>
#include <vector>

namespace multi_camera_people_tracking {
namespace {

const simaai::neat::Tensor* first_tensor_impl(const simaai::neat::Sample& sample) {
  if (sample.kind == simaai::neat::SampleKind::Tensor && sample.tensor.has_value()) {
    return &sample.tensor.value();
  }
  for (const auto& field : sample.fields) {
    if (const auto* tensor = first_tensor_impl(field); tensor != nullptr) {
      return tensor;
    }
  }
  return nullptr;
}

} // namespace

std::vector<Detection> filter_person_detections(const std::vector<Detection>& boxes,
                                                int person_class_id) {
  std::vector<Detection> filtered;
  for (const auto& box : boxes) {
    if (box.class_id == person_class_id) {
      filtered.push_back(box);
    }
  }
  return filtered;
}

OptiViewTrackingPayload
make_optiview_tracking_detection(const std::vector<TrackedDetection>& tracked) {
  OptiViewTrackingPayload payload;
  payload.objects.reserve(tracked.size());
  payload.labels.reserve(tracked.size());
  for (std::size_t index = 0; index < tracked.size(); ++index) {
    const auto& det = tracked[index];
    sima_examples::OptiViewObject object;
    object.x = static_cast<int>(det.x1);
    object.y = static_cast<int>(det.y1);
    object.w = static_cast<int>(std::max(0.0f, det.x2 - det.x1));
    object.h = static_cast<int>(std::max(0.0f, det.y2 - det.y1));
    object.score = det.score;
    object.class_id = static_cast<int>(index);
    payload.objects.push_back(object);
    payload.labels.push_back("Track ID: " + std::to_string(det.track_id));
  }
  return payload;
}

std::vector<std::uint8_t> extract_bbox_payload(const simaai::neat::Sample& sample) {
  std::vector<std::uint8_t> payload;
  std::string error;
  if (!sima_examples::extract_bbox_payload(sample, payload, error)) {
    return {};
  }
  return payload;
}

std::vector<Detection> parse_bbox_payload(const std::vector<std::uint8_t>& payload, int img_w,
                                          int img_h) {
  const auto parsed = objdet::parse_boxes_lenient(payload, img_w, img_h, 0);
  std::vector<Detection> out;
  out.reserve(parsed.size());
  for (const auto& box : parsed) {
    out.push_back(Detection{
        box.x1,
        box.y1,
        box.x2,
        box.y2,
        box.score,
        box.class_id,
    });
  }
  return out;
}

cv::Mat tensor_rgb_from_sample(const simaai::neat::Sample& sample) {
  const auto* tensor = first_tensor_impl(sample);
  if (tensor == nullptr) {
    throw std::runtime_error("no tensor payload found in decoded RTSP sample");
  }
  return tensor->to_cv_mat_copy(simaai::neat::ImageSpec::PixelFormat::RGB);
}

} // namespace multi_camera_people_tracking

File: examples/tracking/multi-camera-people-detection-and-tracking-optiview/python/utils/sample_utils.py
"""Sample and payload helpers for the Python multi-camera example."""

from __future__ import annotations

import struct
from typing import Any

from .tracker import TrackedDetection


def filter_person_detections(boxes: list[dict], person_class_id: int = 0) -> list[dict]:
    return [box for box in boxes if int(box.get("class_id", -1)) == int(person_class_id)]


def make_optiview_tracking_detection(
    pyneat: Any,
    tracked: list[TrackedDetection],
) -> tuple[list[Any], list[str]]:
    objects: list[Any] = []
    labels: list[str] = []
    for index, det in enumerate(tracked):
        obj = pyneat.OptiViewObject()
        obj.x = int(det.x1)
        obj.y = int(det.y1)
        obj.w = int(max(0.0, det.x2 - det.x1))
        obj.h = int(max(0.0, det.y2 - det.y1))
        obj.score = float(det.score)
        obj.class_id = index
        objects.append(obj)
        labels.append(f"Track ID: {det.track_id}")
    return objects, labels


def iter_tensors(pyneat: Any, sample: Any):
    if getattr(sample, "kind", None) == pyneat.SampleKind.Tensor and getattr(sample, "tensor", None) is not None:
        yield sample.tensor
    for tensor in getattr(sample, "tensors", []):
        yield tensor
    for field in getattr(sample, "fields", []):
        yield from iter_tensors(pyneat, field)


def first_tensor(pyneat: Any, sample: Any) -> Any:
    if sample is None:
        return None
    if getattr(sample, "kind", None) == pyneat.SampleKind.Tensor and getattr(sample, "tensor", None) is not None:
        return sample.tensor
    for tensor in iter_tensors(pyneat, sample):
        return tensor
    return None


def extract_bbox_payload(pyneat: Any, sample: Any) -> bytes | None:
    for tensor in iter_tensors(pyneat, sample):
        try:
            payload = tensor.copy_payload_bytes()
        except Exception:
            continue
        if payload:
            return payload
    return None


def tensor_bgr_from_sample(runtime, sample: Any):
    tensor = first_tensor(runtime.pyneat, sample)
    if tensor is None:
        raise RuntimeError("no tensor payload found in decoded RTSP sample")

    if tensor.is_nv12():
        width = int(tensor.width() if callable(tensor.width) else tensor.width)
        height = int(tensor.height() if callable(tensor.height) else tensor.height)
        payload = runtime.np.frombuffer(tensor.copy_payload_bytes(), dtype=runtime.np.uint8)
        expected = width * height * 3 // 2
        if payload.size < expected:
            raise RuntimeError(f"NV12 payload too small: {payload.size} < {expected}")
        nv12 = payload[:expected].reshape((height * 3 // 2, width))
        bgr = runtime.cv2.cvtColor(nv12, runtime.cv2.COLOR_YUV2BGR_NV12)
        rgb = runtime.cv2.cvtColor(bgr, runtime.cv2.COLOR_BGR2RGB)
        return runtime.np.ascontiguousarray(rgb)

    if tensor.is_i420():
        width = int(tensor.width() if callable(tensor.width) else tensor.width)
        height = int(tensor.height() if callable(tensor.height) else tensor.height)
        payload = runtime.np.frombuffer(tensor.copy_payload_bytes(), dtype=runtime.np.uint8)
        expected = width * height * 3 // 2
        if payload.size < expected:
            raise RuntimeError(f"I420 payload too small: {payload.size} < {expected}")
        i420 = payload[:expected].reshape((height * 3 // 2, width))
        bgr = runtime.cv2.cvtColor(i420, runtime.cv2.COLOR_YUV2BGR_I420)
        rgb = runtime.cv2.cvtColor(bgr, runtime.cv2.COLOR_BGR2RGB)
        return runtime.np.ascontiguousarray(rgb)

    if hasattr(tensor, "to_numpy"):
        try:
            arr = runtime.np.asarray(tensor.to_numpy(copy=True))
        except Exception:
            arr = None
        if arr is not None:
            if arr.ndim == 4 and arr.shape[0] == 1:
                arr = arr[0]
            if arr.ndim != 3:
                raise RuntimeError(f"unexpected decoded tensor shape: {arr.shape}")
            if arr.dtype != runtime.np.uint8:
                arr = runtime.np.clip(arr, 0, 255).astype(runtime.np.uint8)
            return runtime.np.ascontiguousarray(arr)

    if hasattr(tensor, "to_cv_mat_copy"):
        return tensor.to_cv_mat_copy(runtime.pyneat.PixelFormat.RGB)

    raise RuntimeError("decoded sample tensor does not expose a supported conversion API")


def parse_bbox_payload(payload: bytes | None, img_w: int, img_h: int) -> list[dict]:
    if not payload or len(payload) < 4:
        return []
    count = min(struct.unpack_from("<I", payload, 0)[0], (len(payload) - 4) // 24)
    boxes: list[dict] = []
    offset = 4
    for _ in range(count):
        x, y, w, h, score, class_id = struct.unpack_from("<iiiifi", payload, offset)
        offset += 24
        x1 = max(0.0, min(float(img_w), float(x)))
        y1 = max(0.0, min(float(img_h), float(y)))
        x2 = max(0.0, min(float(img_w), float(x + w)))
        y2 = max(0.0, min(float(img_h), float(y + h)))
        if x2 <= x1 or y2 <= y1:
            continue
        boxes.append(
            {
                "x1": x1,
                "y1": y1,
                "x2": x2,
                "y2": y2,
                "score": float(score),
                "class_id": int(class_id),
            }
        )
    return boxes
